### PR TITLE
fix: flaky test `4byte setting does not try to get contract method name from 4byte when the setting is off`

### LINKED ISSUE #21494: Fix flaky test "does not try to get contract method name from 4byte when the setting is off"
Failed build: https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/100024/workflows/433aa9de-f5b8-495d-abbc-80034c4db6e3/jobs/3723710

### PR DESCRIPTION
**Pull request by seaona**
## **Description**
This test is failing as it's trying to look for the `Deposit initiated` element in the test dapp and it doesn't appear in time.

Screenshot from 2024-10-02 12-21-16

We shouldn't care if the test dapp sets the value to Deposit initiated into its div element, as long as the popup is open (which it does). This removes any potential race condition on the test dapp side

There are several things to fix/improve around the 2 specs for 4byte, so I took the opportunity to fix those too.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27560?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/21494

## **Manual testing steps**

1. Check ci
2. Run test locally

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: test/e2e/tests/settings/4byte-directory.spec.js [hunk 1]
```diff
@@ -83,28 +78,23 @@ describe('4byte setting', function () {
...
           css: '.confirm-page-container-summary__action__name',
           text: contractInteraction,
         });
-        // We add a delay here to wait for any potential UI changes
-        await driver.delay(veryLargeDelayMs);
-        // css text-transform: uppercase is applied to the text
-        assert.equal(
-          await actionElement.getText(),
-          contractInteraction.toUpperCase(),
-        );
```
Comment (seaona) [inline test/e2e/tests/settings/4byte-directory.spec.js:107]: this way of asserting can cause race conditions, when the element it's present but not the value we expected. This is an anti-pattern, that needs to be switch by waiting for a value (text) we want. This way we can safely remove the delay

File: test/e2e/tests/settings/4byte-directory.spec.js [hunk 2]
```diff
@@ -83,28 +78,23 @@ describe('4byte setting', function () {
         await openDapp(driver, contractAddress);
 
         // wait for deployed contract, calls and confirms a contract method where ETH is sent
-        await driver.findClickableElement('#depositButton');
```
Comment (seaona) [inline test/e2e/tests/settings/4byte-directory.spec.js:86]: not needed

File: test/e2e/tests/settings/4byte-directory.spec.js [hunk 3]
```diff
@@ -83,28 +78,23 @@ describe('4byte setting', function () {
...
         await driver.clickElement('#depositButton');
 
-        await driver.waitForSelector({
-          css: 'span',
-          text: 'Deposit initiated',
-        });
-
-        await driver.waitUntilXWindowHandles(3);
```
Comment (seaona) [inline test/e2e/tests/settings/4byte-directory.spec.js:94]: not needed

File: test/e2e/tests/settings/4byte-directory.spec.js [hunk 4]
```diff
@@ -27,27 +26,23 @@ describe('4byte setting', function () {
...
           smartContract,
         );
-        await unlockWallet(driver);
+        await logInWithBalanceValidation(driver);
 
         // deploy contract
         await openDapp(driver, contractAddress);
 
         // wait for deployed contract, calls and confirms a contract method where ETH is sent
-        await driver.delay(largeDelayMs);
```
Comment (seaona) [inline test/e2e/tests/settings/4byte-directory.spec.js:36]: if we make sure we login with balance validation, we can safely remove this delay

File: test/e2e/tests/settings/4byte-directory.spec.js [hunk 5]
```diff
@@ -27,27 +26,23 @@ describe('4byte setting', function () {
...
         await driver.clickElement('#depositButton');
 
-        await driver.waitForSelector({
-          css: 'span',
-          text: 'Deposit initiated',
-        });
```
Comment (seaona) [inline test/e2e/tests/settings/4byte-directory.spec.js:42]: we don't care if the test dapp loads this value in time or not

File: test/e2e/tests/settings/4byte-directory.spec.js [hunk 6]
```diff
@@ -27,27 +25,23 @@ describe('4byte setting', function () {
...
-
-        await driver.waitUntilXWindowHandles(3);
         await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
-        const actionElement = await driver.waitForSelector({
-          css: '.confirm-page-container-summary__action__name',
+        await driver.waitForSelector({
+          tag: 'span',
           text: 'Deposit',
         });
-        assert.equal(await actionElement.getText(), 'DEPOSIT');
```
Comment (seaona) [inline test/e2e/tests/settings/4byte-directory.spec.js:50]: asserting an element by it's inner value to match an expect value, opens the door to race conditions. We should find the element by its value directly. I've also added the 2 conditions we want (deposit is there, and contract interaction is not there)

**Review by DDDDDanica — APPROVED** [review]

**Review by legobeat — APPROVED** [review]

**Review by chloeYue — APPROVED** [review]
LGTM! Thanks for the fix !

**Pull request merged**